rust/driver/src/lib.rs

```diff
@@ -42,255 +46,319 @@ pub unsafe extern "C" fn get_tx_context_zeroed(_context: *mut ffi::c_void) -> ev
-            },
-            stack.len(),
-            if memory.is_empty() {
-                ptr::null_mut()
-            } else {
-                memory.as_mut_ptr()
-            },
-            memory.len(),
-            if last_call_result_data.is_empty() {
-                ptr::null_mut()
-            } else {
-                last_call_result_data.as_mut_ptr()
-            },
-            last_call_result_data.len(),
-            steps,
-        )
+impl Drop for SteppableInstance {
+    fn drop(&mut self) {
+        let destroy = self.0.destroy.unwrap();
+        unsafe { destroy(self.0) };
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-pub fn run_steppable_with_null_context(
-    host: &evmc_host_interface,
-    revision: Revision,
-    message: &evmc_message,
-    code: &[u8],
-    status: evmc_step_status_code,
-    pc: u64,
-    gas_refunds: i64,
-    stack: &mut [Uint256],
-    memory: &mut [u8],
-    last_call_result_data: &mut [u8],
-    steps: i32,
-) -> evmc_step_result {
-    // SAFETY:
-    // All pointer are valid since they are created from references except for `context` which is
-    // allowed to be null.
-    unsafe {
-        run_steppable_raw(
-            host,
-            ptr::null_mut(),
-            revision,
-            message,
-            if code.is_empty() {
-                ptr::null()
-            } else {
-                code.as_ptr()
-            },
-            code.len(),
-            status,
-            pc,
-            gas_refunds,
-            if stack.is_empty() {
-                ptr::null_mut()
-            } else {
-                stack.as_mut_ptr()
-            },
-            stack.len(),
-            if memory.is_empty() {
-                ptr::null_mut()
-            } else {
-                memory.as_mut_ptr()
-            },
-            memory.len(),
-            if last_call_result_data.is_empty() {
-                ptr::null_mut()
-            } else {
-                last_call_result_data.as_mut_ptr()
-            },
-            last_call_result_data.len(),
-            steps,
-        )
+impl SteppableInstance {
+    /// # Safety
+    ///
+    /// All pointers must be valid, except for `context` which can be null if the
+    /// `evmc_host_interface` accepts null pointers as context.
+    #[allow(clippy::too_many_arguments)]
+    pub unsafe fn run_steppable_raw(
+        &mut self,
+        host: *const evmc_host_interface,
+        context: *mut ffi::c_void,
+        revision: Revision,
+        message: *const evmc_message,
+        code: *const u8,
+        code_len: usize,
+        status: evmc_step_status_code,
+        pc: u64,
+        gas_refunds: i64,
+        stack: *mut Uint256,
+        stack_len: usize,
+        memory: *mut u8,
+        memory_len: usize,
+        last_call_result_data: *mut u8,
+        last_call_result_data_len: usize,
+        steps: i32,
+    ) -> evmc_step_result {
+        let step_n = self.step_n.unwrap();
+
+        unsafe {
```

Comment:
if the function is unsafe, do we  need to add unsafe inside? 
can we remove one of the two?

Reply:
No we don't. I'll remove it.